# Coursera IBM Data Science - Capstone Project - Battle of the Neighborhoods

### Please note: If you are viewing this on GITHUB folium maps will not display correctly.Please use the following link in IBM Watson instead: https://dataplatform.cloud.ibm.com/analytics/notebooks/v2/478fa4ee-291a-4a9c-bab1-94d44aa0a381/view?access_token=73679c1c80bda09f1ba518be6abcf173b7ebf2a0f840ac7c59e240a3ef6fc3c4

## Introduction / Business Problem: 
### A client has hired me to help their company find a new location for their 5 star seafood restaurant business. They are trying to decide between opening the new business in Manhattan, NY or Paris France. Their preference is to first narrow down the city by researching market saturation. After that, they would like to identifying the neighborhood that is best suited for the restaurant. The ideal neighborhood will have an avg amount of restaurants, and few seafood restaurants.

## Data: 
### In this project, I will utilize the Foursquare API combined with geographical data to pull in the total restaurants and seafood types for each city. After determining the city, I will use neighborhood data to narrow down the best location. If needed, the NY data will be taken from here: https://cocl.us/new_york_dataset and the Paris dataset will be scraped from : https://en.wikipedia.org/wiki/Arrondissements_of_Paris

### Library imports

In [1]:
import numpy as np # library to handle data in a vectorized manner
!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
!pip install matplotlib
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
import urllib.request # Import library to open URL's
from bs4 import BeautifulSoup # Import Beautiful Soup package to scrape webpage
from pandas.io.json import json_normalize
!pip install folium
import folium
print('Libraries imported.')

     |████████████████████████████████| 102kB 9.8MB/s eta 0:00:01
Libraries imported.


### Use geopy to get latitude and longitude for Manhattan

In [2]:
addressMan = 'Manhattan, NY'

geolocatorMan = Nominatim(user_agent="Man_explorer")
locationMan = geolocatorMan.geocode(addressMan)
latitudeMan = locationMan.latitude
longitudeMan = locationMan.longitude
print('The geographical coordinates of Manhattan are {}, {}.'.format(latitudeMan, longitudeMan))

The geographical coordinates of Manhattan are 40.7896239, -73.9598939.


### There is code in the following cells that is not shown because it displays Foursquare ID and Secret. I will first pass a query to Foursqaure to get the total # of restaurants within 1,000 meters of the Manhattan coordinates to determine market saturation. I will then pass the same criteria to see how many Seafood restaurants are in the same area.

In [3]:
# The code was removed by Watson Studio for sharing.

In [4]:
# The code was removed by Watson Studio for sharing.

'https://api.foursquare.com/v2/venues/explore?categoryId=4d4b7105d754a06374d81259&client_id=OJXT5VI2BYL2RGJV0E3FTUUMP2VQ12GWKEM52W5CMZX41ILZ&client_secret=W4IBWPNUOPPDAFFYGT255I3XESKRXI5MZD22ISC3BWKTAQ10&v=20200801&ll=40.7896239,-73.9598939&radius=1000&limit=1000'

In [5]:
resultsMan = requests.get(url).json()
resultsMan

{'meta': {'code': 200, 'requestId': '5f3298b0af3e194d4077b9a3'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Central Park',
  'headerFullLocation': 'Central Park, New York',
  'headerLocationGranularity': 'neighborhood',
  'query': 'food',
  'totalResults': 62,
  'suggestedBounds': {'ne': {'lat': 40.798623909000014,
    'lng': -73.94802882768683},
   'sw': {'lat': 40.78062389099999, 'lng': -73.97175897231317}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5328a418498e4be839240b62',
       'name': 'Da Capo',
       'location': {'address': '1392 Madison Ave',
        'crossStreet': 'btwn E 96th & E 97th St',
        'lat': 40.78767941353

### Using the totalResults row = 62, we know that there are 62 restaurants within 1,000 meters of the search area. Now, of those 62, how many are Seafood? Again, this cell is protected since it contains my Secret code.

In [6]:
# The code was removed by Watson Studio for sharing.

'https://api.foursquare.com/v2/venues/explore?categoryId=4bf58dd8d48988d1ce941735&client_id=OJXT5VI2BYL2RGJV0E3FTUUMP2VQ12GWKEM52W5CMZX41ILZ&client_secret=W4IBWPNUOPPDAFFYGT255I3XESKRXI5MZD22ISC3BWKTAQ10&v=20200801&ll=40.7896239,-73.9598939&radius=1000&limit=1000'

In [7]:
resultsManSea = requests.get(url).json()
resultsManSea

{'meta': {'code': 200, 'requestId': '5f32983982da01237eed0ed3'},
 'response': {'headerLocation': 'Central Park',
  'headerFullLocation': 'Central Park, New York',
  'headerLocationGranularity': 'neighborhood',
  'query': 'seafood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 40.798623909000014,
    'lng': -73.94802882768683},
   'sw': {'lat': 40.78062389099999, 'lng': -73.97175897231317}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4a862614f964a520960020e3',
       'name': 'Island Restaurant',
       'location': {'address': '1305 Madison Ave',
        'crossStreet': 'at E 92nd St.',
        'lat': 40.78500527012526,
        'lng': -73.9557891850224,
        'labeledLatLngs': [{'label': 'display',
          'lat': 40.78500527012526,
          'lng': -73.

### Using the totalResults row, we now know that there are 5 other seafood locations in the target area. Now...lets compare that with Paris. Let's get the coordinates using Geopy.

In [103]:
addressPrs = 'Paris, FR'

geolocatorPrs = Nominatim(user_agent="Prs_explorer")
locationPrs = geolocatorPrs.geocode(addressPrs)
latitudePrs = locationPrs.latitude
longitudePrs = locationPrs.longitude
print('The geographical coordinates of Paris are {}, {}.'.format(latitudePrs, longitudePrs))

The geographical coordinates of Paris are 48.8566969, 2.3514616.


### Send coordinates to Foursquare for all restaurants in 1,000 meters (hidden cell below)

In [104]:
# The code was removed by Watson Studio for sharing.

'https://api.foursquare.com/v2/venues/explore?categoryId=4d4b7105d754a06374d81259&client_id=OJXT5VI2BYL2RGJV0E3FTUUMP2VQ12GWKEM52W5CMZX41ILZ&client_secret=W4IBWPNUOPPDAFFYGT255I3XESKRXI5MZD22ISC3BWKTAQ10&v=20200801&ll=48.8566969,2.3514616&radius=1000&limit=1000'

In [105]:
resultsPrs = requests.get(urlPrs).json()
resultsPrs

{'meta': {'code': 200, 'requestId': '5f32b036e70bb116fe4bb5dd'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Hôtel-de-Ville',
  'headerFullLocation': 'Hôtel-de-Ville, Paris',
  'headerLocationGranularity': 'neighborhood',
  'query': 'food',
  'totalResults': 249,
  'suggestedBounds': {'ne': {'lat': 48.865696909000015,
    'lng': 2.36511503696949},
   'sw': {'lat': 48.84769689099999, 'lng': 2.33780816303051}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54351b59498e123ee4afc7ff',
       'name': "L'Alsacien",
       'location': {'address': '6 rue Saint-Bon',
        'lat': 48.85827502747949,
        'lng': 2.3503806119494897,
        'labeledLatLngs': [{'label': 'display',
  

### Wow, we can see that there are 249 restaurants in the area! Lets take a look now to see how many of those are seafood. The following cell is hidden.

In [106]:
# The code was removed by Watson Studio for sharing.

'https://api.foursquare.com/v2/venues/explore?categoryId=4bf58dd8d48988d1ce941735&client_id=OJXT5VI2BYL2RGJV0E3FTUUMP2VQ12GWKEM52W5CMZX41ILZ&client_secret=W4IBWPNUOPPDAFFYGT255I3XESKRXI5MZD22ISC3BWKTAQ10&v=20200801&ll=48.8566969,2.3514616&radius=1000&limit=1000'

In [107]:
resultsPrsSea = requests.get(url).json()
resultsPrsSea

{'meta': {'code': 200, 'requestId': '5f32adb311597434c53dc2ce'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Hôtel-de-Ville',
  'headerFullLocation': 'Hôtel-de-Ville, Paris',
  'headerLocationGranularity': 'neighborhood',
  'query': 'seafood',
  'totalResults': 22,
  'suggestedBounds': {'ne': {'lat': 48.865696909000015,
    'lng': 2.36511503696949},
   'sw': {'lat': 48.84769689099999, 'lng': 2.33780816303051}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5471c7d5498e7ddb1b9178ff',
       'name': 'Les Pinces',
       'location': {'address': '29 rue du Bourg Tibourg',
        'lat': 48.85755185997607,
        'lng': 2.3563175494829465,
        'labeledLatLngs': [{'label': 'di

### Now we can see from the totalResults row that 22 restaurants out of 249 are seafood. What is interesting is both cities have about the same percentage of seafood resaturants compared to total. Paris has 9% and Manhattan has 8%

### Due to the heavy saturation of restaurants in Paris, we now feel comfortable in saying that Manhattan is the better place to open a new restaurant. Also, only 5 restaurants in the area are seafood, so the competition is very small comparitively speaking.

### Now that we have narrowed it down to Manhattan, its time to zero in on the recommended neighborhood.

### Download New York Data

In [13]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


### Open the JSON and create the dataframe columns

In [14]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)
neighborhoods_data = newyork_data['features']
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 
# instantiate the dataframe
neighborhoodsMan = pd.DataFrame(columns=column_names)

### Loop through dataset to fill the dataframe

In [15]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoodsMan = neighborhoodsMan.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)
neighborhoodsMan

Borough               Neighborhood   Latitude  Longitude
0            Bronx                  Wakefield  40.894705 -73.847201
1            Bronx                 Co-op City  40.874294 -73.829939
2            Bronx                Eastchester  40.887556 -73.827806
3            Bronx                  Fieldston  40.895437 -73.905643
4            Bronx                  Riverdale  40.890834 -73.912585
5            Bronx                Kingsbridge  40.881687 -73.902818
6        Manhattan                Marble Hill  40.876551 -73.910660
7            Bronx                   Woodlawn  40.898273 -73.867315
8            Bronx                    Norwood  40.877224 -73.879391
9            Bronx             Williamsbridge  40.881039 -73.857446
10           Bronx                 Baychester  40.866858 -73.835798
11           Bronx             Pelham Parkway  40.857413 -73.854756
12           Bronx                City Island  40.847247 -73.786488
13           Bronx               Bedford Park  40.870185 -73.885512
14           Bronx         University Heights  40.855727 -73.910416
15           Bronx             Morris Heights  40.847898 -73.919672
16           Bronx                    Fordham  40.860997 -73.896427
17           Bronx               East Tremont  40.842696 -73.887356
18           Bronx                 West Farms  40.839475 -73.877745
19           Bronx               High  Bridge  40.836623 -73.926102
20           Bronx                    Melrose  40.819754 -73.909422
21           Bronx                 Mott Haven  40.806239 -73.916100
22           Bronx                Port Morris  40.801664 -73.913221
23           Bronx                   Longwood  40.815099 -73.895788
24           Bronx                Hunts Point  40.809730 -73.883315
25           Bronx                 Morrisania  40.823592 -73.901506
26           Bronx                  Soundview  40.821012 -73.865746
27           Bronx               Clason Point  40.806551 -73.854144
28           Bronx                Throgs Neck  40.815109 -73.816350
29           Bronx               Country Club  40.844246 -73.824099
30           Bronx                Parkchester  40.837938 -73.856003
31           Bronx         Westchester Square  40.840619 -73.842194
32           Bronx                   Van Nest  40.843608 -73.866299
33           Bronx                Morris Park  40.847549 -73.850402
34           Bronx                    Belmont  40.857277 -73.888452
35           Bronx             Spuyten Duyvil  40.881395 -73.917190
36           Bronx            North Riverdale  40.908543 -73.904531
37           Bronx                 Pelham Bay  40.850641 -73.832074
38           Bronx              Schuylerville  40.826580 -73.826203
39           Bronx             Edgewater Park  40.821986 -73.813885
40           Bronx                Castle Hill  40.819014 -73.848027
41           Bronx                  Olinville  40.871371 -73.863324
42           Bronx             Pelham Gardens  40.862966 -73.841612
43           Bronx                  Concourse  40.834284 -73.915589
44           Bronx                  Unionport  40.829774 -73.850535
45           Bronx                   Edenwald  40.884561 -73.848083
46        Brooklyn                  Bay Ridge  40.625801 -74.030621
47        Brooklyn                Bensonhurst  40.611009 -73.995180
48        Brooklyn                Sunset Park  40.645103 -74.010316
49        Brooklyn                 Greenpoint  40.730201 -73.954241
50        Brooklyn                  Gravesend  40.595260 -73.973471
51        Brooklyn             Brighton Beach  40.576825 -73.965094
52        Brooklyn             Sheepshead Bay  40.586890 -73.943186
53        Brooklyn          Manhattan Terrace  40.614433 -73.957438
54        Brooklyn                   Flatbush  40.636326 -73.958401
55        Brooklyn              Crown Heights  40.670829 -73.943291
56        Brooklyn              East Flatbush  40.641718 -73.936103
57        Brooklyn                 Kensington  40.642382 -73.980421

In [16]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoodsMan['Borough'].unique()),
        neighborhoodsMan.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


### Select only Manhattan neighborhoods and reset index

In [17]:
df_Man = neighborhoodsMan[neighborhoodsMan['Borough'].str.contains("Manhattan")]
df_Man = df_Man.reset_index(drop=True) # Reset index
df_Man

Borough         Neighborhood   Latitude  Longitude
0   Manhattan          Marble Hill  40.876551 -73.910660
1   Manhattan            Chinatown  40.715618 -73.994279
2   Manhattan   Washington Heights  40.851903 -73.936900
3   Manhattan               Inwood  40.867684 -73.921210
4   Manhattan     Hamilton Heights  40.823604 -73.949688
5   Manhattan       Manhattanville  40.816934 -73.957385
6   Manhattan       Central Harlem  40.815976 -73.943211
7   Manhattan          East Harlem  40.792249 -73.944182
8   Manhattan      Upper East Side  40.775639 -73.960508
9   Manhattan            Yorkville  40.775930 -73.947118
10  Manhattan           Lenox Hill  40.768113 -73.958860
11  Manhattan     Roosevelt Island  40.762160 -73.949168
12  Manhattan      Upper West Side  40.787658 -73.977059
13  Manhattan       Lincoln Square  40.773529 -73.985338
14  Manhattan              Clinton  40.759101 -73.996119
15  Manhattan              Midtown  40.754691 -73.981669
16  Manhattan          Murray Hill  40.748303 -73.978332
17  Manhattan              Chelsea  40.744035 -74.003116
18  Manhattan    Greenwich Village  40.726933 -73.999914
19  Manhattan         East Village  40.727847 -73.982226
20  Manhattan      Lower East Side  40.717807 -73.980890
21  Manhattan              Tribeca  40.721522 -74.010683
22  Manhattan         Little Italy  40.719324 -73.997305
23  Manhattan                 Soho  40.722184 -74.000657
24  Manhattan         West Village  40.734434 -74.006180
25  Manhattan     Manhattan Valley  40.797307 -73.964286
26  Manhattan  Morningside Heights  40.808000 -73.963896
27  Manhattan             Gramercy  40.737210 -73.981376
28  Manhattan    Battery Park City  40.711932 -74.016869
29  Manhattan   Financial District  40.707107 -74.010665
30  Manhattan        Carnegie Hill  40.782683 -73.953256
31  Manhattan                 Noho  40.723259 -73.988434
32  Manhattan         Civic Center  40.715229 -74.005415
33  Manhattan        Midtown South  40.748510 -73.988713
34  Manhattan         Sutton Place  40.760280 -73.963556
35  Manhattan           Turtle Bay  40.752042 -73.967708
36  Manhattan           Tudor City  40.746917 -73.971219
37  Manhattan      Stuyvesant Town  40.731000 -73.974052
38  Manhattan             Flatiron  40.739673 -73.990947
39  Manhattan         Hudson Yards  40.756658 -74.000111

### Time to visualize the Manhattan neighborhoods, we will use the latitude and longitude variables that we set earlier

In [18]:
# create map of Toronto using latitude and longitude values
map_manhattan = folium.Map(location=[latitudeMan, longitudeMan], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_Man['Latitude'], df_Man['Longitude'], df_Man['Borough'], df_Man['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  

map_manhattan

### Now, we will run back out to Foursquare to pull back in our Manhattan data

In [19]:
# The code was removed by Watson Studio for sharing.

'https://api.foursquare.com/v2/venues/explore?categoryId=4d4b7105d754a06374d81259&client_id=OJXT5VI2BYL2RGJV0E3FTUUMP2VQ12GWKEM52W5CMZX41ILZ&client_secret=W4IBWPNUOPPDAFFYGT255I3XESKRXI5MZD22ISC3BWKTAQ10&v=20200801&ll=40.7896239,-73.9598939&radius=1000&limit=1000'

### Check out the results, ensure the total is the same

In [20]:
resultsMan = requests.get(url).json()
resultsMan

{'meta': {'code': 200, 'requestId': '5f32986188ec7f202b2c1a0a'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Central Park',
  'headerFullLocation': 'Central Park, New York',
  'headerLocationGranularity': 'neighborhood',
  'query': 'food',
  'totalResults': 62,
  'suggestedBounds': {'ne': {'lat': 40.798623909000014,
    'lng': -73.94802882768683},
   'sw': {'lat': 40.78062389099999, 'lng': -73.97175897231317}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5328a418498e4be839240b62',
       'name': 'Da Capo',
       'location': {'address': '1392 Madison Ave',
        'crossStreet': 'btwn E 96th & E 97th St',
        'lat': 40.78767941353

### Cleanup and extract data to a dataframe

In [21]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [22]:
restaurants = resultsMan['response']['groups'][0]['items']
    
nearby_restaurants = json_normalize(restaurants) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_restaurants =nearby_restaurants.loc[:, filtered_columns]

# filter the category for each row
nearby_restaurants['venue.categories'] = nearby_restaurants.apply(get_category_type, axis=1)

# clean columns
nearby_restaurants.columns = [col.split(".")[-1] for col in nearby_restaurants.columns]

nearby_restaurants

name  \
0                                             Da Capo   
1                                    Russ & Daughters   
2                                      Bluestone Lane   
3                                  Paola's Restaurant   
4                                Earl's Beer & Cheese   
5                                        Table d'Hote   
6                                     Sammy's Gourmet   
7                                     Sarabeth's East   
8                           Marinara Pizza Upper East   
9                               Bonjour Crêpes & Wine   
10                            Marco Polo Pizza & Deli   
11                             Chipotle Mexican Grill   
12                                           Pascalou   
13                                          Mila Cafe   
14                                          Felo Deli   
15                                           Tre Otto   
16                                              Noglu   
17                            Lexington Pizza Parlour   
18                                         Champignon   
19                                      Café Sabarsky   
20                                            Sfoglia   
21                                     Joy Burger Bar   
22                                     Frenchy Coffee   
23                                     Lex Restaurant   
24                           Absolute Thai Restaurant   
25                               Little Luzzo's Pizza   
26                              Famous Famiglia Pizza   
27                                         Bella Luna   
28      Amy’s Bread At Museum Of The City Of New York   
29                                      Superior Cafe   
30                             Corrado Bread & Pastry   
31                                           Bawarchi   
32                                      EL Paso Truck   
33                                            La Shuk   
34                                         The Wright   
35                                  Gina La Fornarina   
36                                           Pro Thai   
37                                     Domino's Pizza   
38                                    Yura on Madison   
39                                   Lolita's Kitchen   
40                                    Maoz Vegetarian   
41                                    Lexington Grill   
42                                 feta bar and grill   
43                                Lupita's Restaurant   
44  Chalsty's Cafe at The Museum of the City of Ne...   
45                              Coffee Cart - 96th St   
46                                Daniello's Pizzeria   
47                        El Cafe at Museo del Barrio   
48                              Herm's Box Food Truck   
49                                   Roger Deli House   
50                           New Amsterdam Burger Bar   
51                        Trattoria iL Gusto Wine Bar   
52                        Tattoria Pescepasta Quattro   
53                               Tu Casa Deli Grocery   
54                                Sun Deli & Pizzeria   
55                                      Hot Dog stand   
56                                   105 Madison Deli   
57                                    Cafe Fledermaus   
58                                        Fruit Stand   
59                                        103 Express   
60                                           Gan Asia   
61                              Lalu Deli And Grocery   

                       categories        lat        lng  
0                            Café  40.787679 -73.953899  
1               Kosher Restaurant  40.785332 -73.957481  
2                            Café  40.783757 -73.958526  
3                      Restaurant  40.784655 -73.955869  
4             American Restaurant  40.787331 -73.951725  
5               French Restaurant  40.784492 -73.955712  
6                   Deli / Bodega  40.788138 -73.95

### Now lets get restaurants by neighborhood 

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?categoryId=4d4b7105d754a06374d81259&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [24]:
manhattan_venues = getNearbyVenues(names=df_Man['Neighborhood'],
                                   latitudes=df_Man['Latitude'],
                                   longitudes=df_Man['Longitude']
                                  )
manhattan_venues

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


Neighborhood   Latitude  Longitude  \
0             Marble Hill  40.876551 -73.910660   
1             Marble Hill  40.876551 -73.910660   
2             Marble Hill  40.876551 -73.910660   
3             Marble Hill  40.876551 -73.910660   
4             Marble Hill  40.876551 -73.910660   
5             Marble Hill  40.876551 -73.910660   
6             Marble Hill  40.876551 -73.910660   
7             Marble Hill  40.876551 -73.910660   
8             Marble Hill  40.876551 -73.910660   
9             Marble Hill  40.876551 -73.910660   
10            Marble Hill  40.876551 -73.910660   
11            Marble Hill  40.876551 -73.910660   
12            Marble Hill  40.876551 -73.910660   
13            Marble Hill  40.876551 -73.910660   
14              Chinatown  40.715618 -73.994279   
15              Chinatown  40.715618 -73.994279   
16              Chinatown  40.715618 -73.994279   
17              Chinatown  40.715618 -73.994279   
18              Chinatown  40.715618 -73.994279   
19              Chinatown  40.715618 -73.994279   
20              Chinatown  40.715618 -73.994279   
21              Chinatown  40.715618 -73.994279   
22              Chinatown  40.715618 -73.994279   
23              Chinatown  40.715618 -73.994279   
24              Chinatown  40.715618 -73.994279   
25              Chinatown  40.715618 -73.994279   
26              Chinatown  40.715618 -73.994279   
27              Chinatown  40.715618 -73.994279   
28              Chinatown  40.715618 -73.994279   
29              Chinatown  40.715618 -73.994279   
30              Chinatown  40.715618 -73.994279   
31              Chinatown  40.715618 -73.994279   
32              Chinatown  40.715618 -73.994279   
33              Chinatown  40.715618 -73.994279   
34              Chinatown  40.715618 -73.994279   
35              Chinatown  40.715618 -73.994279   
36              Chinatown  40.715618 -73.994279   
37              Chinatown  40.715618 -73.994279   
38              Chinatown  40.715618 -73.994279   
39              Chinatown  40.715618 -73.994279   
40              Chinatown  40.715618 -73.994279   
41              Chinatown  40.715618 -73.994279   
42              Chinatown  40.715618 -73.994279   
43              Chinatown  40.715618 -73.994279   
44              Chinatown  40.715618 -73.994279   
45              Chinatown  40.715618 -73.994279   
46              Chinatown  40.715618 -73.994279   
47              Chinatown  40.715618 -73.994279   
48              Chinatown  40.715618 -73.994279   
49              Chinatown  40.715618 -73.994279   
50              Chinatown  40.715618 -73.994279   
51              Chinatown  40.715618 -73.994279   
52              Chinatown  40.715618 -73.994279   
53              Chinatown  40.715618 -73.994279   
54              Chinatown  40.715618 -73.994279   
55              Chinatown  40.715618 -73.994279   
56              Chinatown  40.715618 -73.994279   
57              Chinatown  40.715618 -73.994279   
58              Chinatown  40.715618 -73.994279   
59              Chinatown  40.715618 -73.994279   
60              Chinatown  40.715618 -73.994279   
61              Chinatown  40.715618 -73.994279   
62              Chinatown  40.715618 -73.994279   
63              Chinatown  40.715618 -73.994279   
64              Chinatown  40.715618 -73.994279   
65              Chinatown  40.715618 -73.994279   
66              Chinatown  40.715618 -73.994279   
67              Chinatown  40.715618 -73.994279   
68              Chinatown  40.715618 -73.994279   
69              Chinatown  40.715618 -73.994279   
70              Chinatown  40.715618 -73.994279   
71              Chinatown  40.715618 -73.994279   
72              Chinatown  40.715618 -73.994279   
73              Chinatown  40.715618 -73.994279   
74              Chinatown  40.715618 -73.994279   
75              Chinatown  40.715618 -73.994279   
76              Chinatown  40.715618 -73.994279   
77              Chinatown  40.71561

### Count # of Restaurants by neighborhood

In [25]:
mancount = manhattan_venues.groupby('Neighborhood').count()
mancount = mancount[['Venue Category']]
mancount.columns = ['Restaurant Count']
mancount

Restaurant Count
Neighborhood                         
Battery Park City                  30
Carnegie Hill                      72
Central Harlem                     44
Chelsea                            82
Chinatown                         100
Civic Center                       78
Clinton                           100
East Harlem                        51
East Village                      100
Financial District                100
Flatiron                          100
Gramercy                           49
Greenwich Village                 100
Hamilton Heights                   57
Hudson Yards                       46
Inwood                             49
Lenox Hill                        100
Lincoln Square                     51
Little Italy                      100
Lower East Side                    42
Manhattan Valley                   42
Manhattanville                     44
Marble Hill                        14
Midtown                           100
Midtown South                     100
Morningside Heights                39
Murray Hill                        79
Noho                              100
Roosevelt Island                   12
Soho                              100
Stuyvesant Town                     6
Sutton Place                       71
Tribeca                            66
Tudor City                         83
Turtle Bay                         90
Upper East Side                    77
Upper West Side                    58
Washington Heights                 69
West Village                      100
Yorkville                          86

### Count by # of Seafood Restaurants - first will filter dataframe to Seafood

In [26]:
mansea2 = manhattan_venues[manhattan_venues['Venue Category'].str.contains("Seafood")]
mansea2 = mansea2.reset_index(drop=True) # Reset index
mansea2

Neighborhood   Latitude  Longitude  \
0           Marble Hill  40.876551 -73.910660   
1             Chinatown  40.715618 -73.994279   
2    Washington Heights  40.851903 -73.936900   
3    Washington Heights  40.851903 -73.936900   
4                Inwood  40.867684 -73.921210   
5      Hamilton Heights  40.823604 -73.949688   
6        Manhattanville  40.816934 -73.957385   
7        Manhattanville  40.816934 -73.957385   
8        Manhattanville  40.816934 -73.957385   
9        Central Harlem  40.815976 -73.943211   
10       Central Harlem  40.815976 -73.943211   
11       Central Harlem  40.815976 -73.943211   
12          East Harlem  40.792249 -73.944182   
13      Upper East Side  40.775639 -73.960508   
14           Lenox Hill  40.768113 -73.958860   
15      Upper West Side  40.787658 -73.977059   
16      Upper West Side  40.787658 -73.977059   
17       Lincoln Square  40.773529 -73.985338   
18              Clinton  40.759101 -73.996119   
19              Clinton  40.759101 -73.996119   
20              Midtown  40.754691 -73.981669   
21          Murray Hill  40.748303 -73.978332   
22              Chelsea  40.744035 -74.003116   
23              Chelsea  40.744035 -74.003116   
24              Chelsea  40.744035 -74.003116   
25    Greenwich Village  40.726933 -73.999914   
26    Greenwich Village  40.726933 -73.999914   
27         East Village  40.727847 -73.982226   
28         East Village  40.727847 -73.982226   
29              Tribeca  40.721522 -74.010683   
30         Little Italy  40.719324 -73.997305   
31         Little Italy  40.719324 -73.997305   
32         Little Italy  40.719324 -73.997305   
33         Little Italy  40.719324 -73.997305   
34                 Soho  40.722184 -74.000657   
35                 Soho  40.722184 -74.000657   
36                 Soho  40.722184 -74.000657   
37         West Village  40.734434 -74.006180   
38         West Village  40.734434 -74.006180   
39         West Village  40.734434 -74.006180   
40  Morningside Heights  40.808000 -73.963896   
41    Battery Park City  40.711932 -74.016869   
42   Financial District  40.707107 -74.010665   
43                 Noho  40.723259 -73.988434   
44                 Noho  40.723259 -73.988434   
45         Sutton Place  40.760280 -73.963556   
46           Turtle Bay  40.752042 -73.967708   
47           Turtle Bay  40.752042 -73.967708   
48           Turtle Bay  40.752042 -73.967708   
49           Tudor City  40.746917 -73.971219   
50           Tudor City  40.746917 -73.971219   
51             Flatiron  40.739673 -73.990947   

                                                Venue  Venue Latitude  \
0                               Land & Sea Restaurant       40.877885   
1                                            The Boil       40.719726   
2   Marisco Centro (Seafood Restaurant & Fish Market)       40.851332   
3                    Mid Atlantic - Discount Fish Mkt       40.849893   
4                          El Tina Seafood Restaurant       40.864460   
5                                  Famous Fish Market       40.823648   
6                   El Puerto Seafood and Fish Market       40.813861   
7                                  Seafood Restaurant       40.814053   
8                         Loui Loui Louisiana Seafood       40.814217   
9                                    House Of Seafood       40.811921   
10                LightHouse Fish Market & Restaurant       40.814267   
11                                  Lenox Fish Market       40.813127   
12                                     Brisas Del Mar       40.795171   
13                                       Flex Mussels       40.776337   
14                             The East Pole Fish Bar       40.769315   
15                                    The Mermaid Inn       40.788744   
16                                     Luke's Lobster       40.784147   
17                                     Atlantic Grill       40.772205   
18                        

### Count the Seafood restaurants by neighborhood and drop unwanted columns

In [27]:
sfdbynbhd = mansea2.groupby(['Neighborhood']).count()
sfdbynbhd = sfdbynbhd[['Venue Category']]
sfdbynbhd.columns = ['Seafood Count']
sfdbynbhd

Seafood Count
Neighborhood                      
Battery Park City                1
Central Harlem                   3
Chelsea                          3
Chinatown                        1
Clinton                          2
East Harlem                      1
East Village                     2
Financial District               1
Flatiron                         1
Greenwich Village                2
Hamilton Heights                 1
Inwood                           1
Lenox Hill                       1
Lincoln Square                   1
Little Italy                     4
Manhattanville                   3
Marble Hill                      1
Midtown                          1
Morningside Heights              1
Murray Hill                      1
Noho                             2
Soho                             3
Sutton Place                     1
Tribeca                          1
Tudor City                       2
Turtle Bay                       3
Upper East Side                  1
Upper West Side                  2
Washington Heights               2
West Village                     3

### Join the two dataframes together to show both counts and percent of restaurants that are seafood

In [28]:
Full_Man = pd.merge(mancount, sfdbynbhd, on='Neighborhood', how='left')
Full_Man['Seafood Count'] = Full_Man['Seafood Count'].fillna(0)
Full_Man['Percent'] = Full_Man['Seafood Count']/Full_Man['Restaurant Count']  
Full_Man['Percent'] = Full_Man['Percent'].astype(float).map("{:.1%}".format)
Full_Man

Restaurant Count  Seafood Count Percent
Neighborhood                                                
Battery Park City                  30            1.0    3.3%
Carnegie Hill                      72            0.0    0.0%
Central Harlem                     44            3.0    6.8%
Chelsea                            82            3.0    3.7%
Chinatown                         100            1.0    1.0%
Civic Center                       78            0.0    0.0%
Clinton                           100            2.0    2.0%
East Harlem                        51            1.0    2.0%
East Village                      100            2.0    2.0%
Financial District                100            1.0    1.0%
Flatiron                          100            1.0    1.0%
Gramercy                           49            0.0    0.0%
Greenwich Village                 100            2.0    2.0%
Hamilton Heights                   57            1.0    1.8%
Hudson Yards                       46            0.0    0.0%
Inwood                             49            1.0    2.0%
Lenox Hill                        100            1.0    1.0%
Lincoln Square                     51            1.0    2.0%
Little Italy                      100            4.0    4.0%
Lower East Side                    42            0.0    0.0%
Manhattan Valley                   42            0.0    0.0%
Manhattanville                     44            3.0    6.8%
Marble Hill                        14            1.0    7.1%
Midtown                           100            1.0    1.0%
Midtown South                     100            0.0    0.0%
Morningside Heights                39            1.0    2.6%
Murray Hill                        79            1.0    1.3%
Noho                              100            2.0    2.0%
Roosevelt Island                   12            0.0    0.0%
Soho                              100            3.0    3.0%
Stuyvesant Town                     6            0.0    0.0%
Sutton Place                       71            1.0    1.4%
Tribeca                            66            1.0    1.5%
Tudor City                         83            2.0    2.4%
Turtle Bay                         90            3.0    3.3%
Upper East Side                    77            1.0    1.3%
Upper West Side                    58            2.0    3.4%
Washington Heights                 69            2.0    2.9%
West Village                      100            3.0    3.0%
Yorkville                          86            0.0    0.0%

### Now I will take only the neighborhoods that do not currently have a seafood restaurant, and have more than 30 restaurants and less than 80. I will also be adding back in coordinates so that we can perform so further analysis on locations.

In [37]:
NoSea = Full_Man[Full_Man['Seafood Count']==0]
NoSea = NoSea[NoSea['Restaurant Count']>30]
NoSea = NoSea[NoSea['Restaurant Count']<80]
NoSea = pd.merge(NoSea, df_Man, on='Neighborhood', how='left')
NoSea

Neighborhood  Restaurant Count  Seafood Count Percent    Borough  \
0     Carnegie Hill                72            0.0    0.0%  Manhattan   
1      Civic Center                78            0.0    0.0%  Manhattan   
2          Gramercy                49            0.0    0.0%  Manhattan   
3      Hudson Yards                46            0.0    0.0%  Manhattan   
4   Lower East Side                42            0.0    0.0%  Manhattan   
5  Manhattan Valley                42            0.0    0.0%  Manhattan   

    Latitude  Longitude  
0  40.782683 -73.953256  
1  40.715229 -74.005415  
2  40.737210 -73.981376  
3  40.756658 -74.000111  
4  40.717807 -73.980890  
5  40.797307 -73.964286

### All of these neighborhoods are great candidates. Let's use some machine learning to further differentiate them for the client. First, we'll plot them out on a map.

In [42]:
# create map of Manhattan using latitude and longitude values
map_candidates = folium.Map(location=[latitudeMan, longitudeMan], zoom_start=13)

# add markers to map
for lat, lng, label in zip(NoSea['Latitude'], NoSea['Longitude'], NoSea['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_candidates)  
    
map_candidates

### Pulling in venue data from Foursquare, this time not limited to restaurants only

In [51]:
# The code was removed by Watson Studio for sharing.

'https://api.foursquare.com/v2/venues/explore?&client_id=OJXT5VI2BYL2RGJV0E3FTUUMP2VQ12GWKEM52W5CMZX41ILZ&client_secret=W4IBWPNUOPPDAFFYGT255I3XESKRXI5MZD22ISC3BWKTAQ10&v=20200801&ll=40.7896239,-73.9598939&radius=1000&limit=1000'

In [52]:
resultsManAll = requests.get(urlven).json()
resultsManAll

{'meta': {'code': 200, 'requestId': '5f329ea099578521cc2300b9'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Central Park',
  'headerFullLocation': 'Central Park, New York',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 116,
  'suggestedBounds': {'ne': {'lat': 40.798623909000014,
    'lng': -73.94802882768683},
   'sw': {'lat': 40.78062389099999, 'lng': -73.97175897231317}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4a78425df964a52053e51fe3',
       'name': 'Central Park Tennis Center',
       'location': {'address': 'Central Park West at 96th St',
        'lat': 40.78931319964619,
        'lng': -73.9618624165804

In [53]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### Build Categories

In [56]:
venues = resultsManAll['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name              categories  \
0                      Central Park Tennis Center            Tennis Court   
1                                    North Meadow                    Park   
2                                     East Meadow                   Field   
3            Jacqueline Kennedy Onassis Reservoir               Reservoir   
4                               The Jewish Museum                  Museum   
5                   Central Park - Woodman's Gate                    Park   
6               Central Park - Gate Of All Saints                    Park   
7                                         Da Capo                    Café   
8             Central Park - Wild West Playground              Playground   
9                                Russ & Daughters       Kosher Restaurant   
10                            K&D Wines & Spirits               Wine Shop   
11  Central Park - North Meadow Recreation Center              Playground   
12                  Central Park - 99th & 5th Ave                    Park   
13        Cooper Hewitt Smithsonian Design Museum              Art Museum   
14                   Jackie Onassis Running Track                   Track   
15                         Tarr Family Playground              Playground   
16                    Oldest Tree in Central Park                    Park   
17                                 Bluestone Lane                    Café   
18                                   Central Park                    Park   
19                             Paola's Restaurant              Restaurant   
20                              Five Pillars Yoga             Yoga Studio   
21            Conservatory Garden Center Fountain                Fountain   
22                           The Corner Bookstore               Bookstore   
23                           Earl's Beer & Cheese     American Restaurant   
24                                   Table d'Hote       French Restaurant   
25                                     The Ravine          Scenic Lookout   
26                            Conservatory Garden                  Garden   
27                 Museum of the City of New York          History Museum   
28                         Whole Foods Wine Store               Wine Shop   
29                    Solomon R Guggenheim Museum              Art Museum   
30                             Whole Foods Market           Grocery Store   
31                    Shape House Upper East Side                     Spa   
32         Central Park - North Meadow Fields 1-4          Baseball Field   
33                     Robert Bendheim Playground              Playground   
34                                Sarabeth's East          Breakfast Spot   
35                                The Guthrie Inn            Cocktail Bar   
36                                   Birch Coffee             Coffee Shop   
37                 Central Park - Vanderbilt Gate                   Plaza   
38                                   Trader Joe's           Grocery Store   
39                               Burnett Fountain                Fountain   
40                         Kitchen Arts & Letters               Bookstore   
41                                    Girls' Gate                   Plaza   
42                                     Vinyl Wine               Wine Shop   
43                   John Purroy Mitchel Memorial        Sculpture Garden   
44                         Columbus Natural Foods           Grocery Store   
45                  Central Park West - W 88th St                    Park   
46                  Central Park West - W 86th St                    Park   
47                   Central Park Cherry Blossoms                  Garden   
48                                    Bridle Path                   Trail   
49                                   Neue Galerie              Art Museum   
50                               Guggenheim Store               Gift Shop   
51                      Marinara Pizza Up

### Get venues by neighborhood

In [65]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [66]:
candidate_venues = getNearbyVenues(names=NoSea['Neighborhood'],
                                   latitudes=NoSea['Latitude'],
                                   longitudes=NoSea['Longitude']
                                  )

Carnegie Hill
Civic Center
Gramercy
Hudson Yards
Lower East Side
Manhattan Valley


In [67]:
print(candidate_venues.shape)
candidate_venues

(433, 7)


Neighborhood   Latitude  Longitude  \
0       Carnegie Hill  40.782683 -73.953256   
1       Carnegie Hill  40.782683 -73.953256   
2       Carnegie Hill  40.782683 -73.953256   
3       Carnegie Hill  40.782683 -73.953256   
4       Carnegie Hill  40.782683 -73.953256   
5       Carnegie Hill  40.782683 -73.953256   
6       Carnegie Hill  40.782683 -73.953256   
7       Carnegie Hill  40.782683 -73.953256   
8       Carnegie Hill  40.782683 -73.953256   
9       Carnegie Hill  40.782683 -73.953256   
10      Carnegie Hill  40.782683 -73.953256   
11      Carnegie Hill  40.782683 -73.953256   
12      Carnegie Hill  40.782683 -73.953256   
13      Carnegie Hill  40.782683 -73.953256   
14      Carnegie Hill  40.782683 -73.953256   
15      Carnegie Hill  40.782683 -73.953256   
16      Carnegie Hill  40.782683 -73.953256   
17      Carnegie Hill  40.782683 -73.953256   
18      Carnegie Hill  40.782683 -73.953256   
19      Carnegie Hill  40.782683 -73.953256   
20      Carnegie Hill  40.782683 -73.953256   
21      Carnegie Hill  40.782683 -73.953256   
22      Carnegie Hill  40.782683 -73.953256   
23      Carnegie Hill  40.782683 -73.953256   
24      Carnegie Hill  40.782683 -73.953256   
25      Carnegie Hill  40.782683 -73.953256   
26      Carnegie Hill  40.782683 -73.953256   
27      Carnegie Hill  40.782683 -73.953256   
28      Carnegie Hill  40.782683 -73.953256   
29      Carnegie Hill  40.782683 -73.953256   
30      Carnegie Hill  40.782683 -73.953256   
31      Carnegie Hill  40.782683 -73.953256   
32      Carnegie Hill  40.782683 -73.953256   
33      Carnegie Hill  40.782683 -73.953256   
34      Carnegie Hill  40.782683 -73.953256   
35      Carnegie Hill  40.782683 -73.953256   
36      Carnegie Hill  40.782683 -73.953256   
37      Carnegie Hill  40.782683 -73.953256   
38      Carnegie Hill  40.782683 -73.953256   
39      Carnegie Hill  40.782683 -73.953256   
40      Carnegie Hill  40.782683 -73.953256   
41      Carnegie Hill  40.782683 -73.953256   
42      Carnegie Hill  40.782683 -73.953256   
43      Carnegie Hill  40.782683 -73.953256   
44      Carnegie Hill  40.782683 -73.953256   
45      Carnegie Hill  40.782683 -73.953256   
46      Carnegie Hill  40.782683 -73.953256   
47      Carnegie Hill  40.782683 -73.953256   
48      Carnegie Hill  40.782683 -73.953256   
49      Carnegie Hill  40.782683 -73.953256   
50      Carnegie Hill  40.782683 -73.953256   
51      Carnegie Hill  40.782683 -73.953256   
52      Carnegie Hill  40.782683 -73.953256   
53      Carnegie Hill  40.782683 -73.953256   
54      Carnegie Hill  40.782683 -73.953256   
55      Carnegie Hill  40.782683 -73.953256   
56      Carnegie Hill  40.782683 -73.953256   
57      Carnegie Hill  40.782683 -73.953256   
58      Carnegie Hill  40.782683 -73.953256   
59      Carnegie Hill  40.782683 -73.953256   
60      Carnegie Hill  40.782683 -73.953256   
61      Carnegie Hill  40.782683 -73.953256   
62      Carnegie Hill  40.782683 -73.953256   
63      Carnegie Hill  40.782683 -73.953256   
64      Carnegie Hill  40.782683 -73.953256   
65      Carnegie Hill  40.782683 -73.953256   
66      Carnegie Hill  40.782683 -73.953256   
67      Carnegie Hill  40.782683 -73.953256   
68      Carnegie Hill  40.782683 -73.953256   
69      Carnegie Hill  40.782683 -73.953256   
70      Carnegie Hill  40.782683 -73.953256   
71      Carnegie Hill  40.782683 -73.953256   
72      Carnegie Hill  40.782683 -73.953256   
73      Carnegie Hill  40.782683 -73.953256   
74      Carnegie Hill  40.782683 -73.953256   
75      Carnegie Hill  40.782683 -73.953256   
76      Carnegie Hill  40.782683 -73.953256   
77      Carnegie Hill  40.782683 -73.953256   
78      Carnegie Hill  40.782683 -73.953256   
79      Carnegie Hill  40.782683 -73.953256   
80      Carnegie Hill  40.782683 -73.953256   
81      Carnegie Hill  40.782683 -73.953256   
82      Carnegie Hill  40.782683 -73.953256   
83      Carnegie Hill  40.782683 -73.953256   
84      Carneg

### Unique Categories

In [70]:
print('There are {} unique categories.'.format(len(candidate_venues['Venue Category'].unique())))

There are 144 unique categories.


### Analyze each neighborhood

In [75]:
# one hot encoding
candidates_onehot = pd.get_dummies(candidate_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
candidates_onehot['Neighborhood'] = candidate_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [candidates_onehot.columns[-1]] + list(candidates_onehot.columns[:-1])
candidates_onehot = candidates_onehot[fixed_columns]

candidates_onehot.head()

Neighborhood  American Restaurant  Antique Shop  Argentinian Restaurant  \
0  Carnegie Hill                    0             0                       0   
1  Carnegie Hill                    0             0                       0   
2  Carnegie Hill                    0             0                       0   
3  Carnegie Hill                    0             0                       0   
4  Carnegie Hill                    0             0                       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  Australian Restaurant  Baby Store  Bagel Shop  Bakery  \
0                   0                      0           0           0       0   
1                   0                      0           0           0       0   
2                   0                      0           0           0       0   
3                   0                      0           0           0       0   
4                   0                      0           0           0       0   

   Bank  Bar  Beer Bar  Bike Rental / Bike Share  Boat or Ferry  Bookstore  \
0     0    0         0                         0              0          1   
1     0    0         0                         0              0          0   
2     0    0         0                         0              0          0   
3     0    0         0                         0              0          0   
4     0    0         0                         0              0          0   

   Boxing Gym  Breakfast Spot  Bubble Tea Shop  Building  Burger Joint  \
0           0               0                0         0             0   
1           0               0                0         0             0   
2           0               0                0         0             0   
3           0               0                0         0             0   
4           0               0                0         0             0   

   Burrito Place  Café  Cajun / Creole Restaurant  Camera Store  \
0              0     0                          0             0   
1              0     0                          0             0   
2              0     0                          0             0   
3              0     0                          0             0   
4              0     0                          0             0   

   Caribbean Restaurant  Caucasian Restaurant  Cheese Shop  \
0                     0                     0            0   
1                     0                     0            0   
2                     0                     0            0   
3                     0                     0            0   
4                     0                     0            0   

   Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  Comedy Club  \
0                   0               0             0            0            0   
1                   0               0             0            0            0   
2                   0               0             0            0            0   
3                   0               0             0            0            0   
4                   0               0             0            0            0   

   Community Center  Convenience Store  Cosmetics Shop  Cupcake Shop  \
0                 0                  0               0             0   
1                 0                  0               0             0   
2                 0                  0               0             0   
3                 0                  0               0             0   
4                 1                  0               0             0   

   Dance Studio  Deli /

### Group by neighborhood and get mean of the frequency of occurence

In [77]:
candidates_grouped = candidates_onehot.groupby('Neighborhood').mean().reset_index()
candidates_grouped

Neighborhood  American Restaurant  Antique Shop  \
0     Carnegie Hill             0.011111          0.00   
1      Civic Center             0.020000          0.01   
2          Gramercy             0.043011          0.00   
3      Hudson Yards             0.052632          0.00   
4   Lower East Side             0.000000          0.00   
5  Manhattan Valley             0.000000          0.00   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                0.011111     0.000000    0.011111             0.000000   
1                0.000000     0.000000    0.000000             0.000000   
2                0.000000     0.010753    0.000000             0.000000   
3                0.000000     0.000000    0.000000             0.000000   
4                0.021277     0.042553    0.000000             0.000000   
5                0.000000     0.000000    0.000000             0.021739   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  Baby Store  \
0              0.00            0.000000                   0.00        0.00   
1              0.01            0.000000                   0.01        0.01   
2              0.00            0.000000                   0.00        0.00   
3              0.00            0.017544                   0.00        0.00   
4              0.00            0.000000                   0.00        0.00   
5              0.00            0.000000                   0.00        0.00   

   Bagel Shop    Bakery      Bank       Bar  Beer Bar  \
0    0.011111  0.022222  0.011111  0.011111  0.000000   
1    0.010000  0.020000  0.000000  0.000000  0.000000   
2    0.053763  0.010753  0.010753  0.064516  0.010753   
3    0.000000  0.000000  0.000000  0.000000  0.000000   
4    0.021277  0.042553  0.000000  0.000000  0.000000   
5    0.000000  0.021739  0.000000  0.065217  0.000000   

   Bike Rental / Bike Share  Boat or Ferry  Bookstore  Boxing Gym  \
0                  0.000000       0.000000   0.033333        0.00   
1                  0.000000       0.000000   0.010000        0.01   
2                  0.010753       0.000000   0.000000        0.00   
3                  0.000000       0.017544   0.000000        0.00   
4                  0.000000       0.000000   0.000000        0.00   
5                  0.000000       0.000000   0.000000        0.00   

   Breakfast Spot  Bubble Tea Shop  Building  Burger Joint  Burrito Place  \
0        0.011111         0.000000  0.000000      0.011111           0.00   
1        0.000000         0.000000  0.010000      0.010000           0.01   
2        0.000000         0.000000  0.000000      0.000000           0.00   
3        0.000000         0.000000  0.017544      0.017544           0.00   
4        0.000000         0.021277  0.000000      0.000000           0.00   
5        0.000000         0.021739  0.000000      0.000000           0.00   

       Café  Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  \
0  0.044444                       0.00      0.000000              0.000000   
1  0.020000                       0.01      0.000000              0.000000   
2  0.021505                       0.00      0.000000              0.000000   
3  0.017544                       0.00      0.017544              0.000000   
4  0.042553                       0.00      0.000000              0.000000   
5  0.021739                       0.00      0.000000              0.021739   

   Caucasian Restaurant  Cheese Shop  Chinese Restaurant  Clothing Store  \
0              0.000000     0.000000            0.011111        0.011111   
1              0.000000     0.000000            0.000000        0.010000   
2              0.000000     0.010753            0.000000        0.000000   
3              0.017544     0.000000            0.000000        0.000000   
4              0.000000     0.000000            0.063830        0.000000   
5              0.000000     0.000000            0.000000        0.021739   

   Cocktail Bar  Coffee

### Top 5 most common venues by neighborhood

In [78]:
num_top_venues = 5

for hood in candidates_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = candidates_grouped[candidates_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Carnegie Hill----
               venue  freq
0        Coffee Shop  0.07
1               Café  0.04
2        Yoga Studio  0.03
3     Cosmetics Shop  0.03
4  French Restaurant  0.03


----Civic Center----
                  venue  freq
0           Coffee Shop  0.09
1          Cocktail Bar  0.05
2                 Hotel  0.05
3  Gym / Fitness Center  0.04
4     French Restaurant  0.04


----Gramercy----
                 venue  freq
0                  Bar  0.06
1           Bagel Shop  0.05
2  American Restaurant  0.04
3          Pizza Place  0.04
4          Coffee Shop  0.04


----Hudson Yards----
                  venue  freq
0  Gym / Fitness Center  0.07
1                 Hotel  0.07
2   American Restaurant  0.05
3    Italian Restaurant  0.05
4               Dog Run  0.04


----Lower East Side----
                venue  freq
0  Chinese Restaurant  0.06
1         Coffee Shop  0.04
2              Bakery  0.04
3        Cocktail Bar  0.04
4         Art Gallery  0.04


----Manhattan Valley-

In [79]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Top 10 venues by neighborhood

In [81]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = candidates_grouped['Neighborhood']

for ind in np.arange(candidates_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(candidates_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0     Carnegie Hill           Coffee Shop                  Café   
1      Civic Center           Coffee Shop                 Hotel   
2          Gramercy                   Bar            Bagel Shop   
3      Hudson Yards                 Hotel  Gym / Fitness Center   
4   Lower East Side    Chinese Restaurant                  Café   
5  Manhattan Valley    Mexican Restaurant                   Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Yoga Studio             Bookstore     French Restaurant   
1          Cocktail Bar  Gym / Fitness Center     French Restaurant   
2   American Restaurant    Italian Restaurant           Coffee Shop   
3   American Restaurant    Italian Restaurant                  Park   
4           Pizza Place   Japanese Restaurant           Art Gallery   
5           Coffee Shop           Pizza Place       Thai Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Gym / Fitness Center        Cosmetics Shop           Pizza Place   
1                  Park    Italian Restaurant           Yoga Studio   
2           Pizza Place            Playground    Mexican Restaurant   
3             Nightclub               Dog Run           Coffee Shop   
4          Cocktail Bar           Coffee Shop                Bakery   
5        Farmers Market            Playground        Clothing Store   

  9th Most Common Venue 10th Most Common Venue  
0                   Gym              Wine Shop  
1                   Spa    Sporting Goods Shop  
2        Ice Cream Shop          Grocery Store  
3            Restaurant                    Gym  
4     Convenience Store       Ramen Restaurant  
5   Peruvian Restaurant                   Park

### Cluster neighborhoods

In [86]:
# set number of clusters
kclusters = 3

candidates_grouped_clustering = candidates_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=2).fit(candidates_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 2, 1, 0, 2], dtype=int32)

In [97]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

candidates_merged = NoSea

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
candidates_merged = candidates_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

candidates_merged.head() # check the last columns!

Neighborhood  Restaurant Count  Seafood Count Percent    Borough  \
0    Carnegie Hill                72            0.0    0.0%  Manhattan   
1     Civic Center                78            0.0    0.0%  Manhattan   
2         Gramercy                49            0.0    0.0%  Manhattan   
3     Hudson Yards                46            0.0    0.0%  Manhattan   
4  Lower East Side                42            0.0    0.0%  Manhattan   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  40.782683 -73.953256               0           Coffee Shop   
1  40.715229 -74.005415               0           Coffee Shop   
2  40.737210 -73.981376               1                   Bar   
3  40.756658 -74.000111               2                 Hotel   
4  40.717807 -73.980890               3    Chinese Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café           Yoga Studio             Bookstore   
1                 Hotel          Cocktail Bar  Gym / Fitness Center   
2            Bagel Shop   American Restaurant    Italian Restaurant   
3  Gym / Fitness Center   American Restaurant    Italian Restaurant   
4                  Café           Pizza Place   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     French Restaurant  Gym / Fitness Center        Cosmetics Shop   
1     French Restaurant                  Park    Italian Restaurant   
2           Coffee Shop           Pizza Place            Playground   
3                  Park             Nightclub               Dog Run   
4           Art Gallery          Cocktail Bar           Coffee Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Pizza Place                   Gym              Wine Shop  
1           Yoga Studio                   Spa    Sporting Goods Shop  
2    Mexican Restaurant        Ice Cream Shop          Grocery Store  
3           Coffee Shop            Restaurant                    Gym  
4                Bakery     Convenience Store       Ramen Restaurant

### Examine clusters

In [98]:
candidates_merged.loc[candidates_merged['Cluster Labels'] == 0, candidates_merged.columns[[1] + list(range(5, candidates_merged.shape[1]))]]

Restaurant Count   Latitude  Longitude  Cluster Labels  \
0                72  40.782683 -73.953256               0   
1                78  40.715229 -74.005415               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                  Café           Yoga Studio   
1           Coffee Shop                 Hotel          Cocktail Bar   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0             Bookstore     French Restaurant  Gym / Fitness Center   
1  Gym / Fitness Center     French Restaurant                  Park   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0        Cosmetics Shop           Pizza Place                   Gym   
1    Italian Restaurant           Yoga Studio                   Spa   

  10th Most Common Venue  
0              Wine Shop  
1    Sporting Goods Shop

In [99]:
candidates_merged.loc[candidates_merged['Cluster Labels'] == 1, candidates_merged.columns[[1] + list(range(5, candidates_merged.shape[1]))]]

Restaurant Count  Latitude  Longitude  Cluster Labels  \
2                49  40.73721 -73.981376               1   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
2                   Bar            Bagel Shop   American Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
2    Italian Restaurant           Coffee Shop           Pizza Place   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
2            Playground    Mexican Restaurant        Ice Cream Shop   

  10th Most Common Venue  
2          Grocery Store

In [100]:
candidates_merged.loc[candidates_merged['Cluster Labels'] == 2, candidates_merged.columns[[1] + list(range(5, candidates_merged.shape[1]))]]

Restaurant Count   Latitude  Longitude  Cluster Labels  \
3                46  40.756658 -74.000111               2   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
3                 Hotel  Gym / Fitness Center   American Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
3    Italian Restaurant                  Park             Nightclub   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
3               Dog Run           Coffee Shop            Restaurant   

  10th Most Common Venue  
3                    Gym

In [101]:
candidates_merged.loc[candidates_merged['Cluster Labels'] == 3, candidates_merged.columns[[1] + list(range(5, candidates_merged.shape[1]))]]

Restaurant Count   Latitude  Longitude  Cluster Labels  \
4                42  40.717807  -73.98089               3   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
4    Chinese Restaurant                  Café           Pizza Place   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
4   Japanese Restaurant           Art Gallery          Cocktail Bar   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
4           Coffee Shop                Bakery     Convenience Store   

  10th Most Common Venue  
4       Ramen Restaurant

In [102]:
candidates_merged.loc[candidates_merged['Cluster Labels'] == 4, candidates_merged.columns[[1] + list(range(5, candidates_merged.shape[1]))]]

Restaurant Count   Latitude  Longitude  Cluster Labels  \
5                42  40.797307 -73.964286               4   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
5    Mexican Restaurant                   Bar           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
5           Pizza Place       Thai Restaurant        Farmers Market   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5            Playground        Clothing Store   Peruvian Restaurant   

  10th Most Common Venue  
5                   Park

### Recommendation:
Although all 6 neighborhoods are good candidates for a new seafood restaurant, I am recommending the Lower East Side location. This is due to a good number of restaurants (42) that is not oversaturated. We can also tell that there is a fairly healthy restaurant culture since there are 5 restaurants in the top 10 venues. Also, an Art Gallery and Cocktail bar in the top 10 venues makes a good compliment to an upscale restaurant. Another great reason is this neighborhood does not contain a seafood restaurant at this time.  